Client/Frontend/AuthenticationManager/AuthenticationSettingsViewController.swift

```diff
@@ -147,45 +151,88 @@ class TouchIDSetting: Setting {
         // we can disable interaction of the switch and still handle tap events.
         let control = UISwitch()
         control.onTintColor = UIConstants.ControlTintColor
-        control.on = authInfo?.useTouchID ?? false
+        self.authInfo = KeychainWrapper.authenticationInfo()
+        control.on = authenticationDelegate?.authInfoUseAuthentication ?? false
         control.userInteractionEnabled = false
         switchControl = control
 
         let accessoryContainer = UIView(frame: control.frame)
         accessoryContainer.addSubview(control)
 
-        let gesture = UITapGestureRecognizer(target: self, action: #selector(TouchIDSetting.switchTapped))
+        let gesture = UITapGestureRecognizer(target: self, action: #selector(AuthenticationSetting.switchTapped))
         accessoryContainer.addGestureRecognizer(gesture)
 
         cell.accessoryView = accessoryContainer
     }
 
     @objc private func switchTapped() {
+        self.authInfo = KeychainWrapper.authenticationInfo()
         guard let authInfo = authInfo else {
             logger.error("Authentication info should always be present when modifying Touch ID preference.")
             return
         }
-
-        if authInfo.useTouchID {
+        guard let authenticationDelegate = authenticationDelegate else {
+            logger.error("The Touch ID setting needs to refer to a specific instance.")
+            return
+        }
+        
+        if authenticationDelegate.authInfoUseAuthentication {
             AppAuthenticator.presentAuthenticationUsingInfo(
                 authInfo,
                 touchIDReason: AuthenticationStrings.disableTouchReason,
-                success: self.touchIDSuccess,
+                success: {
+                    self.touchIDSuccess?(self)
+                },
                 cancel: nil,
-                fallback: self.touchIDFallback
+                fallback: {
+                    self.touchIDFallback?(self)
+                }
             )
         } else {
             toggleTouchID(enabled: true)
         }
     }
 
     func toggleTouchID(enabled enabled: Bool) {
-        authInfo?.useTouchID = enabled
+        self.authInfo = KeychainWrapper.authenticationInfo()
+        authenticationDelegate?.authInfoUseAuthentication = enabled
         KeychainWrapper.setAuthenticationInfo(authInfo)
         switchControl?.setOn(enabled, animated: true)
     }
 }
 
+protocol AuthenticationSettingDelegate: class {
```

Comment:
Couple of things:

1. Technically a 'delegate' in Cocoa land is used as a proxy interface to delegate responses and requests to another object. Since this doesn't define any methods this feels more like a regular Swift protocol instead of a delegate.

2. Protocol-based programming and composition over inheritance are definitely the way to go in most cases but I think it feels out of place here. There's a tight coupling between objects that implement this protocol and the setting object assigned to it. I think this will be cleaner/easier to use by defining `AuthenticationForPrivateBrowsingSetting` and `AuthenticationForLoginsSetting` as direct subclasses of `AuthenticationSetting` instead of using this protocol.